mesonbuild/modules/unstable_external_project.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2020 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os, subprocess, shlex
+from pathlib import Path
+
+from . import ExtensionModule, ModuleReturnValue
+from .. import mlog, build
+from ..mesonlib import (MesonException, Popen_safe, MachineChoice,
+                       get_variable_regex, do_replacement)
+from ..interpreterbase import InterpreterObject, InterpreterException, FeatureNew
+from ..interpreterbase import stringArgs, permittedKwargs
+from ..interpreter import DependencyHolder, InstallDir
+from ..compilers.compilers import cflags_mapping, cexe_mapping
+from ..dependencies.base import InternalDependency, PkgConfigDependency
+
+class ExternalProject(InterpreterObject):
+    def __init__(self, interpreter, subdir, project_version, subproject, environment, build_machine, host_machine,
+                 configure_command, configure_options, cross_configure_options, env, verbose):
+        InterpreterObject.__init__(self)
+        self.methods.update({'dependency': self.dependency_method,
+                             })
+
+        self.interpreter = interpreter
+        self.subdir = Path(subdir)
+        self.project_version = project_version
+        self.subproject = subproject
+        self.env = environment
+        self.build_machine = build_machine
+        self.host_machine = host_machine
+        self.configure_command = configure_command
+        self.configure_options = configure_options
+        self.cross_configure_options = cross_configure_options
+        self.verbose = verbose
+        self.user_env = env
+
+        self.name = self.subdir.name
+        self.src_dir = Path(self.env.get_source_dir(), self.subdir)
```

Comment:
This is by no means wrong, but `Path(self.env.get_source_dir()) / self.subdir` would also work (only the first operand has to be a `Path` object).

This is just in case you didn't know already and boils down to personal preference though, so use what you like better.

Reply:
I knew, thanks. To be fair I was lazy and started by simply s/os.path.join/Path/ :)